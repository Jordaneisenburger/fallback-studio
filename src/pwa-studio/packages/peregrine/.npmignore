.eslintrc.js
/lib/**/__{docs,helpers,mocks,tests}__/**
/scripts
