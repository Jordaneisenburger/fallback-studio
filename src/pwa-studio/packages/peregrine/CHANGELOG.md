# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="0.5.0"></a>
# 0.5.0 (2018-08-13)


### Features

* multicasting REST API client ([#164](https://github.com/magento/pwa-studio/issues/164)) ([2852e14](https://github.com/magento/pwa-studio/commit/2852e14)), closes [#140](https://github.com/magento/pwa-studio/issues/140)
* Simulators for async testing ([#24](https://github.com/magento/pwa-studio/issues/24)) ([a380da9](https://github.com/magento/pwa-studio/commit/a380da9))
