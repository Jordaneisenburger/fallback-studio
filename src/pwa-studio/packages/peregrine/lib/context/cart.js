import React, { createContext, useContext, useMemo } from 'react';
import { connect } from 'react-redux';

import actions from '../store/actions/cart/actions';
import * as asyncActions from '../store/actions/cart/asyncActions';
import bindActionCreators from '../util/bindActionCreators';

const CartContext = createContext();

const isCartEmpty = cart =>
    !cart || !cart.details.items || cart.details.items.length === 0;

const CartContextProvider = props => {
    const { actions, asyncActions, cartState, children } = props;

    const derivedCartState = {
        ...cartState,
        isEmpty: isCartEmpty(cartState)
    };

    const cartApi = useMemo(
        () => ({
            actions,
            ...asyncActions
        }),
        [actions, asyncActions]
    );

    const contextValue = useMemo(() => [derivedCartState, cartApi], [
        cartApi,
        derivedCartState
    ]);

    return (
        <CartContext.Provider value={contextValue}>
            {children}
        </CartContext.Provider>
    );
};

const mapStateToProps = ({ cart }) => ({ cartState: cart });

const mapDispatchToProps = dispatch => ({
    actions: bindActionCreators(actions, dispatch),
    asyncActions: bindActionCreators(asyncActions, dispatch)
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CartContextProvider);

export const useCartContext = () => useContext(CartContext);
