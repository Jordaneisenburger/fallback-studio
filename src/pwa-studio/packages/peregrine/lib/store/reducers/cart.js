import { handleActions } from 'redux-actions';

import actions from '../actions/cart';
import checkoutActions from '../actions/checkout';

export const name = 'cart';

export const initialState = {
    addItemError: null,
    cartId: null,
    details: {},
    detailsError: null,
    isLoading: false,
    isUpdatingItem: false,
    isAddingItem: false,
    paymentMethods: [],
    removeItemError: null,
    shippingMethods: [],
    totals: {},
    updateItemError: null
};

const reducerMap = {
    [actions.getCart.receive]: (state, { payload, error }) => {
        if (error) {
            return initialState;
        }

        return {
            ...state,
            cartId: String(payload)
        };
    },
    [actions.getDetails.request]: (state, { payload }) => {
        return {
            ...state,
            cartId: String(payload),
            isLoading: true
        };
    },
    [actions.getDetails.receive]: (state, { payload, error }) => {
        if (error) {
            return {
                ...state,
                detailsError: payload,
                cartId: null,
                isLoading: false
            };
        }

        return {
            ...state,
            ...payload,
            isLoading: false
        };
    },
    [actions.addItem.request]: state => {
        return {
            ...state,
            isAddingItem: true
        };
    },
    [actions.addItem.receive]: (state, { payload, error }) => {
        if (error) {
            return {
                ...state,
                addItemError: payload,
                isAddingItem: false
            };
        }

        return {
            ...state,
            isAddingItem: false
        };
    },
    [actions.updateItem.request]: (state, { payload, error }) => {
        if (error) {
            return initialState;
        }
        return {
            ...state,
            ...payload,
            isUpdatingItem: true
        };
    },
    [actions.updateItem.receive]: (state, { payload, error }) => {
        if (error) {
            return {
                ...state,
                isUpdatingItem: false,
                updateItemError: payload
            };
        }

        // We don't actually have to update any items here
        // because we force a refresh from the server.
        return {
            ...state,
            isUpdatingItem: false
        };
    },
    [actions.removeItem.receive]: (state, { payload, error }) => {
        if (error) {
            return {
                ...initialState,
                removeItemError: payload
            };
        }
        // If we are emptying the cart, perform a reset to prevent
        // a bug where the next item added to cart would have a price of 0
        if (payload.cartItemCount == 1) {
            return initialState;
        }
        return {
            ...state,
            ...payload
        };
    },
    [checkoutActions.order.accept]: () => {
        return initialState;
    },
    [actions.reset]: () => initialState
};

export default handleActions(reducerMap, initialState);
