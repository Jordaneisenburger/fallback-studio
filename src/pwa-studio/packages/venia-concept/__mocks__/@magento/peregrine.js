import React from 'react';
import { BrowserPersistence } from './util/simplePersistence';

const peregrine = jest.requireActual('../../../peregrine/lib');

// re-exports
const RestApi = {
    Magento2: {
        request: jest.fn()
    }
};

const Util = { BrowserPersistence };

// hooks
const useApolloContext = jest.fn(peregrine.useApolloContext);
const useEventListener = jest.fn(peregrine.useEventListener);
const useDropdown = jest.fn(peregrine.useDropdown);
const useQuery = jest.fn(peregrine.useQuery);
const useQueryResult = jest.fn(peregrine.useQueryResult);
const useSearchParam = jest.fn(peregrine.useSearchParam);
const useWindowSize = jest.fn(peregrine.useWindowSize);
const useToasts = jest.fn(peregrine.useToasts);
// components

/**
 * the Price component from @magento/peregrine
 * has browser-specific functionality and cannot
 * currently by rendered in the test environment
 */
const Price = jest.fn().mockReturnValue(<div />);

module.exports = {
    ...peregrine,
    Price,
    RestApi,
    Util,
    useApolloContext,
    useEventListener,
    useDropdown,
    useQuery,
    useQueryResult,
    useSearchParam,
    useWindowSize,

    useToasts
};
