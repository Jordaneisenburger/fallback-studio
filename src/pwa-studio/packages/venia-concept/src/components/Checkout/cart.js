import React from 'react';
import { bool, func, shape, string } from 'prop-types';

import { mergeClasses } from '../../classify';
import CheckoutButton from './checkoutButton';
import defaultClasses from './cart.css';

const Cart = props => {
    const { beginCheckout, ready, submitting } = props;
    const classes = mergeClasses(defaultClasses, props.classes);

    const disabled = submitting || !ready;

    return (
        <div className={classes.root}>
            <CheckoutButton disabled={disabled} onClick={beginCheckout} />
        </div>
    );
};

Cart.propTypes = {
    beginCheckout: func.isRequired,
    classes: shape({
        root: string
    }),
    ready: bool.isRequired,
    submitting: bool.isRequired
};

export default Cart;
