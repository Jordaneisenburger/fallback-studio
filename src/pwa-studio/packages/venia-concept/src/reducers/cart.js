import { handleActions } from 'redux-actions';

import actions from 'src/actions/cart';
import checkoutActions from 'src/actions/checkout';

export const name = 'cart';

export const initialState = {
    details: {},
    guestCartId: null,
    isLoading: false,
    isOptionsDrawerOpen: false,
    isUpdatingItem: false,
    paymentMethods: [],
    shippingMethods: [],
    totals: {}
};

const reducerMap = {
    [actions.getGuestCart.receive]: (state, { payload, error }) => {
        if (error) {
            return initialState;
        }

        return {
            ...state,
            guestCartId: payload
        };
    },
    [actions.getDetails.request]: (state, { payload }) => {
        return {
            ...state,
            guestCartId: payload,
            isLoading: true
        };
    },
    [actions.getDetails.receive]: (state, { payload, error }) => {
        if (error) {
            return {
                ...state,
                isLoading: false,
                guestCartId: null
            };
        }

        return {
            ...state,
            ...payload,
            isLoading: false
        };
    },
    [actions.updateItem.request]: (state, { payload, error }) => {
        if (error) {
            return initialState;
        }
        return {
            ...state,
            ...payload,
            isUpdatingItem: true
        };
    },
    [actions.updateItem.receive]: state => {
        // We don't actually have to update any items here
        // because we force a refresh from the server.
        return {
            ...state,
            isUpdatingItem: false
        };
    },
    [actions.removeItem.receive]: (state, { payload, error }) => {
        if (error) {
            return initialState;
        }
        // If we are emptying the cart, perform a reset to prevent
        // a bug where the next item added to cart would have a price of 0
        if (payload.cartItemCount == 1) {
            return initialState;
        }
        return {
            ...state,
            ...payload
        };
    },
    [actions.openOptionsDrawer]: state => {
        return {
            ...state,
            isOptionsDrawerOpen: true
        };
    },
    [actions.closeOptionsDrawer]: state => {
        return {
            ...state,
            isOptionsDrawerOpen: false
        };
    },
    [checkoutActions.order.accept]: () => {
        return initialState;
    }
};

export default handleActions(reducerMap, initialState);
