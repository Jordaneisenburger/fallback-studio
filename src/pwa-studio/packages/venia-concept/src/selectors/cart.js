export const isEmptyCartVisible = ({ cart, checkout: { step } }) =>
    step === 'cart' && (!cart.details.items || cart.details.items.length === 0);

export const isMiniCartMaskOpen = ({ checkout: { step } }) => step === 'form';
