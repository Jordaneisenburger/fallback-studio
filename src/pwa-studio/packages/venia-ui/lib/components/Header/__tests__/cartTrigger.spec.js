import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import { useCartContext } from '@magento/peregrine/lib/context/cart';

import CartTrigger from '../cartTrigger';

jest.mock('@magento/peregrine/lib/context/cart', () => {
    const state = {
        details: {}
    };
    const api = {
        getCartDetails: jest.fn(),
        toggleCart: jest.fn()
    };

    const useCartContext = jest.fn(() => [state, api]);

    return { useCartContext };
});

const classes = {
    root: 'a'
};

test('Cart icon svg has no fill when cart is empty', () => {
    const [cartState, cartApi] = useCartContext();
    useCartContext.mockReturnValueOnce([
        {
            ...cartState,
            details: {
                items_qty: 0
            }
        },
        {
            ...cartApi
        }
    ]);

    const component = createTestInstance(<CartTrigger classes={classes} />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('Cart icon svg has fill and correct value when cart contains items', () => {
    const [cartState, cartApi] = useCartContext();
    useCartContext.mockReturnValueOnce([
        {
            ...cartState,
            details: {
                items_qty: 10
            }
        },
        {
            ...cartApi
        }
    ]);

    const component = createTestInstance(<CartTrigger classes={classes} />);

    expect(component.toJSON()).toMatchSnapshot();
});
