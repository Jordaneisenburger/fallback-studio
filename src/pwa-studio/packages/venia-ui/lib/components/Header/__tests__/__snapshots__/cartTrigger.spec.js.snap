// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart icon svg has fill and correct value when cart contains items 1`] = `
<button
  aria-label="Toggle mini cart. You have 10 items in your cart."
  className="a"
  onClick={[Function]}
>
  <span>
    <svg
      fill="rgb(var(--venia-text))"
      height="24"
      stroke="rgb(var(--venia-text))"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="9"
        cy="21"
        r="1"
      />
      <circle
        cx="20"
        cy="21"
        r="1"
      />
      <path
        d="M1 1h4l2.68 13.39a2 2 0 0 0 2 1.61h9.72a2 2 0 0 0 2-1.61L23 6H6"
      />
    </svg>
  </span>
  <span>
    10
  </span>
</button>
`;

exports[`Cart icon svg has no fill when cart is empty 1`] = `
<button
  aria-label="Toggle mini cart. You have 0 items in your cart."
  className="a"
  onClick={[Function]}
>
  <span>
    <svg
      fill="none"
      height="24"
      stroke="rgb(var(--venia-text))"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="9"
        cy="21"
        r="1"
      />
      <circle
        cx="20"
        cy="21"
        r="1"
      />
      <path
        d="M1 1h4l2.68 13.39a2 2 0 0 0 2 1.61h9.72a2 2 0 0 0 2-1.61L23 6H6"
      />
    </svg>
  </span>
</button>
`;
